operators/tokenizer/bpe_utils.hpp

```diff
@@ -100,28 +100,30 @@ class TokenWithRegularExp {
   }
 
  private:
-  std::u32string_view TryMatch() {
-    // python pattern:
-    // 's|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+
-
-    // 's|'t|'re|'ve|'m|'ll|'d|
-    // Note: the sequencial of the following if should not be switched, which follows the python regex's syntax
-    if ((m_text[0] == U'\'') && (m_text.size() > 1)) {
-      if ((m_text[1] == U's') || (m_text[1] == U't') ||
-          (m_text[1] == U'm') || (m_text[1] == U'd')) {
-        std::u32string_view res = m_text.substr(0, 2);
-        m_text = m_text.substr(2);
-        return res;
-      }
-
-      if (m_text.size() > 2) {
-        if (((m_text[1] == U'r') && (m_text[2] == U'e')) ||
-            ((m_text[1] == U'v') && (m_text[2] == U'e')) ||
-            ((m_text[1] == U'l') && (m_text[2] == U'l'))) {
-          std::u32string_view res = m_text.substr(0, 3);
-          m_text = m_text.substr(3);
+  std::u32string_view TryMatch(bool skipApostrophes) {
+    if (!skipApostrophes) {
```

Comment:
can you copy the Python implementation here for better understanding?

Reply:
sounds good - issue is Python implementation has same regex as us but from manual testing it works differently: for example given `r"""<\|startoftext\|>|<\|endoftext\|>|'s|'t|'re|'ve|'m|'ll|'d|[\p{L}]+|[\p{N}]|[^\s\p{L}\p{N}]+"""`, the `'s|'t|'re|'ve|'m|'ll|'d` part splits text like "_you're_" to ["_you_", "_'re_"] given our old implementation, but Python imp gives ["_you_", "_'_", "_re_"].

`skipApostrophes` is intended to bring parity with the regex compiling for Python. I can update the comment, that is not as clear as this explanation.